server/asset/eth/eth_test.go

```diff
@@ -225,31 +212,27 @@ func TestDecodeCoinID(t *testing.T) {
 }
 
 func TestRun(t *testing.T) {
-	// TODO: Test all paths.
 	ctx, cancel := context.WithCancel(context.Background())
-	header1 := &types.Header{Number: big.NewInt(1)}
-	block1 := types.NewBlockWithHeader(header1)
-	blockHash1 := block1.Hash()
-	node := &testNode{
-		bestBlkHash: blockHash1,
-		blk:         block1,
-	}
-
 	backend := unconnectedETH(tLogger, new(config))
 	ch := backend.BlockChannel(1)
+	node := &testNode{
+		bestHdr: &types.Header{Number: big.NewInt(0)},
+	}
 	backend.node = node
+	backend.hashCache.prime(ctx, node)
+	*node = testNode{
+		bestHdr: &types.Header{
+			Number: big.NewInt(1),
+		},
+	}
 	go func() {
 		<-ch
 		cancel()
 	}()
 	backend.run(ctx)
-	backend.blockCache.mtx.Lock()
-	best := backend.blockCache.best
-	backend.blockCache.mtx.Unlock()
-	if best.hash != blockHash1 {
-		t.Fatalf("want header hash %x but got %x", blockHash1, best.hash)
-	}
-	cancel()
+	// This timing out indicates failure. run should call a block update
+	// after the pollInterval and we have the listening function cancel the
+	// context above.
```

Comment:
Hmm. The `context.Context` doesn't have a timeout on it and I don't see how you differentiate a timeout from a success. Do you mean the default test timeout of 10 minutes which results in a panic?

Reply:
yes, a panic. Can add a shorter timeout and real failure.

Reply:
Shortened.